// @kinecosystem/kin-sdk-node/scripts/bin/blockchain/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HorizonErrorList = {
    BAD_REQUEST: 'bad_request',
    BEFORE_HISTORY: 'before_history',
    FORBIDDEN: 'forbidden',
    NOT_ACCEPTABLE: 'not_acceptable',
    NOT_FOUND: 'not_found',
    NOT_IMPLEMENTED: 'not_implemented',
    RATE_LIMIT_EXCEEDED: 'rate_limit_exceeded',
    SERVER_OVER_CAPACITY: 'server_over_capacity',
    STALE_HISTORY: 'stale_history',
    TIMEOUT: 'timeout',
    TRANSACTION_MALFORMED: 'transaction_malformed',
    TRANSACTION_FAILED: 'transaction_failed',
    UNSUPPORTED_MEDIA_TYPE: 'unsupported_media_type',
    INTERNAL_SERVER_ERROR: 'server_error'
};
exports.TransactionErrorList = {
    FAILED: 'tx_failed',
    TOO_EARLY: 'tx_too_early',
    TOO_LATE: 'tx_too_late',
    MISSING_OPERATION: 'tx_missing_operation',
    BAD_SEQUENCE: 'tx_bad_seq',
    BAD_AUTH: 'tx_bad_auth',
    INSUFFICIENT_BALANCE: 'tx_insufficient_balance',
    NO_ACCOUNT: 'tx_no_source_account',
    INSUFFICIENT_FEE: 'tx_insufficient_fee',
    BAD_AUTH_EXTRA: 'tx_bad_auth_extra',
    INTERNAL_ERROR: 'tx_internal_error'
};
exports.OperationResultCode = {
    INNER: 'op_inner',
    BAD_AUTH: 'op_bad_auth',
    NO_ACCOUNT: 'op_no_source_account',
    NOT_SUPPORTED: 'op_not_supported',
    SUCCESS: 'op_success'
};
exports.CreateAccountResultCode = {
    SUCCESS: 'op_success',
    MALFORMED: 'op_malformed',
    UNDERFUNDED: 'op_underfunded',
    LOW_RESERVE: 'op_low_reserve',
    ACCOUNT_EXISTS: 'op_already_exists'
};
exports.PaymentResultCode = {
    SUCCESS: 'op_success',
    MALFORMED: 'op_malformed',
    UNDERFUNDED: 'op_underfunded',
    SRC_NO_TRUST: 'op_src_no_trust',
    SRC_NOT_AUTHORIZED: 'op_src_not_authorized',
    NO_DESTINATION: 'op_no_destination',
    NO_TRUST: 'op_no_trust',
    NOT_AUTHORIZED: 'op_not_authorized',
    LINE_FULL: 'op_line_full',
    NO_ISSUER: 'op_no_issuer'
};
exports.ChangeTrustResultCode = {
    SUCCESS: 'op_success',
    MALFORMED: 'op_malformed',
    NO_ISSUER: 'op_no_issuer',
    LOW_RESERVE: 'op_low_reserve',
    INVALID_LIMIT: 'op_invalid_limit'
};
