// @kinecosystem/kin-sdk-node/scripts/bin/blockchain/keyPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const kin_sdk_1 = require("@kinecosystem/kin-sdk");
const crypto = require("crypto");
class KeyPair {
    get seed() {
        return this._keypair.secret();
    }
    get publicAddress() {
        return this._keypair.publicKey();
    }
    constructor(seed, seedBuffer) {
        if (seed) {
            this._keypair = kin_sdk_1.Keypair.fromSecret(seed);
        }
        else if (seedBuffer) {
            this._keypair = kin_sdk_1.Keypair.fromRawEd25519Seed(seedBuffer);
        }
        else {
            this._keypair = kin_sdk_1.Keypair.random();
        }
    }
    sign(data) {
        this._keypair.sign(data);
    }
    static fromSeed(seed) {
        return new KeyPair(seed, undefined);
    }
    static generate() {
        return new KeyPair(undefined, undefined);
    }
    static generateHDSeed(baseSeed, salt) {
        const hash = crypto.createHash('sha256').update(baseSeed + salt).digest();
        return new KeyPair(undefined, hash);
    }
    static addressFromSeed(seed) {
        return kin_sdk_1.Keypair.fromSecret(seed).publicKey();
    }
}
exports.KeyPair = KeyPair;
